File: strings/04_sort_vowels/sort_vowels.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <cmath>
#include <stdexcept>

using namespace std;


class Solution {
public:
    string sortVowels(const string& s) {
        if (!checkConstraints(s)) {
            throw std::runtime_error("Constraints violated");
        }

        string t(s);
        vector<int> vowel_idx;
        map<char, int> vowel_qnt;

        for (int i = 0; i < s.size(); i++) {
            if ((s[i] == 'a' || s[i] == 'e' || s[i] == 'i' || s[i] == 'o' || s[i] == 'u') || 
                (s[i] == 'A' || s[i] == 'E' || s[i] == 'I' || s[i] == 'O' || s[i] == 'U')) {
                vowel_idx.push_back(i);
                vowel_qnt[s[i]]++;
            }
        }
        if (vowel_idx.size() != 0) {
            int id = 0;
            for (auto& p : vowel_qnt) {
                for (int i = 0; i < p.second; i++) {
                    t[vowel_idx[id]] = p.first;
                    id++;
                }
            }
        } else {
            return s;
        }
        return t;
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const string& s) {
        if (s.size() < 1 || s.size() > pow(10, 5)) return false;

        int A_dec = 65;
        int a_dec = 97;
        int alpha_len = 26;
        for (const char& ch : s) {
            int ascii = static_cast<int>(ch);
            if (!((ascii >= A_dec && ascii < A_dec + alpha_len) || (ascii >= a_dec && ascii < a_dec + alpha_len))) {
                return false;
            }
        }
        return true;
    }
};

File: strings/04_sort_vowels/sort_vowels.py
from sortedcontainers import SortedDict


class Solution:
    def sort_vowels(self, s: str) -> str:
        if not self._check_constraints(s):
            raise ValueError("Constraints violated")

        t = list(s)
        vowel_idx = []
        vowel_qnt = SortedDict()
        for i in range(len(s)):
            if (s[i] == 'a' or s[i] == 'e' or s[i] == 'i' or s[i] == 'o' or s[i] == 'u') or (s[i] == 'A' or s[i] == 'E' or s[i] == 'I' or s[i] == 'O' or s[i] == 'U'):
                vowel_idx.append(i)
                vowel_qnt[s[i]] += 1

        if len(vowel_idx) != 0:
            id = 0
            for key, value in vowel_qnt.items():
                for i in range(value):
                    t[vowel_idx[id]] = key
                    id += 1
            else:
                return s
        return ''.join(t)

    def _check_constraints(self, s: str) -> bool:
        n = len(s)
        if n < 1 or n > 10**5:
            return False

        A_dec = 65
        a_dec = 97
        alpha_len = 26
        for c in s:
            ascii = int(c)
            if not ((ascii >= A_dec and ascii < A_dec + alpha_len) or (ascii >= a_dec and ascii < a_dec + alpha_len)):
                return False

        return True
